### ISSUE TITLE
Calls made to the Execution Manager do not return data

### ISSUE THREAD

**Issue by snario** [ethereum-optimism/optimism#474]
Presently it is impossible to get any return data data from transactions or calls made that go through the `ExecutionManager` such as revert reasons or other kinds of data `geth` may want to use. We've made a [pull request](https://github.com/ethereum-optimism/contracts/pull/342) in the past to address this but decided it was not the right way to go since a change was queued up (and is now [implemented](https://github.com/ethereum-optimism/solidity/pull/22)) to the Solidity compiler that modifies how the execution manager is interacted with. This issue is to document the work to be done to revive that PR based on the new work in the Solidity compiler.

**Comment by platocrat** [comment]
To add to this issue, not getting any return data from calls affects our integration tests, e.g. our tutorials and code examples for our users.

I think the desired behavior, if a transaction fails, is that L2 geth should get the revert reason returned from the contract instead of throwing an estimate gas error.

**Comment by snario** [comment]
Another previous issue: https://github.com/ethereum-optimism/go-ethereum/issues/288

**Comment by platocrat** [comment]
A hacker, @popcorndao, also experienced this issue.

He also noticed that after setting gasPrice to 0 in his tests and config, he had to the `gasLimit` to `8900000` in his tests, otherwise, they would fail.

**Comment by snario** [comment]
@platocrat / @popcorndao it would be helpful to demonstrate a reproduction scenario for what you're describing above. By "this issue", I'm assuming you mean that gas estimation is failing?

**Comment by platocrat** [comment]
To be clear, there are 2 separate issues with integration testing that are related to this GitHub issue:

## 1
As I understand it, `eth_estimateGas` is not failing.
After talking about this issue with @K-Ho, the following error:
```
Error: cannot estimate gas; transaction may fail or may require a manual gas limit
```
is thrown whenever an L2 contract call reverts. Transactions will successfully revert, but we don't get the revert reason that we should be getting. Instead, we get the error shown above.
This issue is confirmed by observations of the tests for the [Waffle-ERC20-Example](https://github.com/ethereum-optimism/Waffle-ERC20-Example) repository

## 2
From having run variations of tests for the [Truffle-ERC20-Example](https://github.com/ethereum-optimism/Truffle-ERC20-Example), there is a _different_ error that is thrown when attempting to deploy a contract. The error thrown when deploying a contract with Truffle is something like:
```
Error: while migrating ERC20: invalid transaction: insufficient funds for gas * price + value
```

This error is thrown even after setting gasPrice to 0 in the network config and in the deployment and contract calls.

--------

All that being said, it seems to me that these two separate errors are related

**Comment by snario** [comment]
Thanks for the summary @platocrat. It seems like (1) exactly describes this issue, and a new issue with reproduction steps should be opened for (2), do you agree?

**Comment by platocrat** [comment]
> Thanks for the summary @platocrat. It seems like (1) exactly describes this issue, and a new issue with reproduction steps should be opened for (2), do you agree?

I agree. I'll open a new issue for the Truffle tests

**Comment by platocrat** [comment]
@snario any updates on this?

**Comment by TransmissionsDev** [comment]
This is a pretty major devex pain right now 😅

**Comment by snario** [comment]
Fixed by https://github.com/ethereum-optimism/optimism/pull/643

**Comment by snario** [comment]
This will be available in the 0.3.0 release schedule for mainnext next week.

**Comment by TransmissionsDev** [comment]
WOOHOO
